server/src/main/java/com/linecorp/centraldogma/server/metadata/MetadataService.java

```diff
@@ -673,128 +706,127 @@ public CompletableFuture<Revision> updateWriteQuota(
-    private CompletableFuture<Collection<Permission>> findPermissions0(String projectName, String repoName,
-                                                                       User user) {
+    private CompletableFuture<RepositoryRole> findRepositoryRole0(String projectName, String repoName,
+                                                                  User user) {
         requireNonNull(projectName, "projectName");
         requireNonNull(repoName, "repoName");
         requireNonNull(user, "user");
 
         return getProject(projectName).thenApply(metadata -> {
             final RepositoryMetadata repositoryMetadata = metadata.repo(repoName);
-            final Member member = metadata.memberOrDefault(user.id(), null);
+            final Roles roles = repositoryMetadata.roles();
+            final RepositoryRole userRepositoryRole = roles.users().get(user.id());
 
-            // If the member is guest.
-            if (member == null) {
-                return repositoryMetadata.perRolePermissions().guest();
-            }
-            final Collection<Permission> p = repositoryMetadata.perUserPermissions().get(member.id());
-            if (p != null) {
-                return p;
-            }
-            return findPerRolePermissions(repositoryMetadata, member.role());
+            final Member projectUser = metadata.memberOrDefault(user.id(), null);
+            final ProjectRole projectRole = projectUser != null ? projectUser.role() : ProjectRole.GUEST;
+            return repositoryRole(roles, userRepositoryRole, projectRole);
         });
     }
 
-    private static Collection<Permission> findPerRolePermissions(RepositoryMetadata repositoryMetadata,
-                                                                 ProjectRole role) {
-        switch (role) {
-            case OWNER:
-                return repositoryMetadata.perRolePermissions().owner();
-            case MEMBER:
-                return repositoryMetadata.perRolePermissions().member();
-            default:
-                return repositoryMetadata.perRolePermissions().guest();
+    @Nullable
+    private static RepositoryRole repositoryRole(Roles roles, @Nullable RepositoryRole repositoryRole,
+                                                 @Nullable ProjectRole projectRole) {
+        if (repositoryRole == RepositoryRole.ADMIN || projectRole == ProjectRole.OWNER) {
+            return RepositoryRole.ADMIN;
         }
+
+        final RepositoryRole memberOrGuestRole;
+        if (projectRole == ProjectRole.MEMBER) {
+            memberOrGuestRole = roles.projectRoles().member();
+        } else {
+            memberOrGuestRole = roles.projectRoles().guest();
+        }
+
+        if (repositoryRole == null) {
+            return memberOrGuestRole;
+        }
+        if (memberOrGuestRole == null) {
+            return repositoryRole;
+        }
+        if (memberOrGuestRole == RepositoryRole.WRITE || repositoryRole == RepositoryRole.WRITE) {
```

Comment:
Let's say `a-member-user` is granted to `READ` role and `projects.member` is granted to `WRITE`. Is `WRITE` given for `a-member-user`?
```json
{
  "repos": {
    "my-repo": {
      "name": "my-repo",
      "roles": {
        "users": {
          "a-member-user": "READ"
        }
        "projects": {
          "member": "WRITE",
          ...
       }
      }
    }
  }
}
```

Reply:
> Is WRITE given for a-member-user?

That is correct, as similar to how GitHub handles it.
The role with the highest authority will take precedence.

Reply:
That's a little surprising - I was expecting the repository role to override the project role.
If that's how Github is doing it, then I guess the current approach is fine.